tests/task/test_api_sampler.py

```diff
@@ -0,0 +1,172 @@
+    return False
+
+class MockGameState(GameState):
+  def __init__(self):
+    # pylint: disable=super-init-not-called
+    pass
+
+class TestTaskAPI(unittest.TestCase):
+
+  def test_operators(self):
+    # pylint: disable=unsupported-binary-operation,invalid-unary-operand-type
+
+    mock_gs = MockGameState()
+
+    # AND (&), OR (|), NOT (~), IMPLY (>>)
+    task1 = Success() & Failure() & Success()
+    self.assertFalse(task1(mock_gs))
+
+    task2 = Success() | Failure() | Success()
+    self.assertTrue(task2(mock_gs))
+
+    task3 = Success() &  ~ Failure() & Success()
+    self.assertTrue(task3(mock_gs))
+
+    task4 = Success() >> Success()
+    self.assertTrue(task4(mock_gs))
+
+    task5 = Success() >> ~ Success()
+    self.assertFalse(task5(mock_gs))
+
+    task6 = (Failure() >> Failure()) & Success()
+    self.assertTrue(task6(mock_gs))
+
+  def test_predicate_name(self):
+
+    success = Success()
+    failure = Failure()
+    fake_task = FakePredicate(Group([2]), 1, Item.Hat, Action.Melee)
+    combination = (success & ~ (failure | fake_task)) | (failure >> fake_task)
+
+    self.assertEqual(combination.name,
+      "OR(AND(Success,NOT(OR(Failure,FakePredicate_(2,)_1_Hat_Melee))),"
+      "IMPLY(Failure->FakePredicate_(2,)_1_Hat_Melee))")
+
+  def test_team_helper(self):
+    # TODO(kywch): This test is true now but may change later.
+
+    config = ScriptedAgentTestConfig()
+    env = nmmo.Env(config)
+    env.reset()
+
+    team_helper = TeamHelper(list(range(1, config.PLAYER_N+1)), len(config.PLAYERS))
+
+    # agents' population should match team_helper team id
+    for ent_id, ent in env.realm.players.items():
+      # pylint: disable=protected-access
+      self.assertEqual(team_helper._ent_to_team[ent_id], ent.population)
+
+  def test_team_assignment(self):
+    team =  Group([1, 2, 8, 9], "TeamFoo")
+
+    self.assertEqual(team.name, 'TeamFoo')
+    self.assertEqual(team[2].name, "TeamFoo.2")
+    self.assertEqual(team[2], (8,))
+
+    # don't allow member of one-member team
+    self.assertEqual(team[2][0].name, team[2].name)
+
+  def test_random_task_sampler(self):
+    rand_sampler = sampler.RandomTaskSampler()
+
+    rand_sampler.add_task_spec(Success, [[Group([1]), Group([3])]])
+    rand_sampler.add_task_spec(Failure, [[Group([2]), Group([1,3])]])
+    rand_sampler.add_task_spec(FakePredicate, [
+      [Group([1]), Group([2]), Group([1,2]), Group([3]), Group([1,3])],
+      [Item.Hat, Item.Top, Item.Bottom],
+      [1, 5, 10],
+      [0.1, 0.2, 0.3, 0.4]
+    ])
+
+    rand_sampler.sample(max_clauses=4, max_clause_size=3, not_p=0.5)
+
+  def test_completed_tasks_in_info(self):
+    config = ScriptedAgentTestConfig()
+    env = TaskEnv(config)
+
+    # some team helper maybe necessary
+    team_helper = TeamHelper(range(1, config.PLAYER_N+1), len(config.PLAYERS))
+
+    fake_task = FakePredicate(team_helper.left_team(3), Item.Hat, 1, 0.1)
+    task_assignment = MultiTask(
```

Comment:
In the multi-agent setting, would there be a case that we don't use MultiTask?

Reply:
Refactored out MultiTask.